File: 543-diameter-of-binary-tree/543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def calculateHeight(self, root: Optional[TreeNode], diameter: list) -> int:
        if not root: return 0
        leftHeight = self.calculateHeight(root.left, diameter)
        rightHeight = self.calculateHeight(root.right, diameter)
        diameter[0] = max(diameter[0], leftHeight + rightHeight)
        return 1 + max(leftHeight, rightHeight)
    
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        diameter = [0]
        self.calculateHeight(root, diameter)
        return diameter[0]

File: 543-diameter-of-binary-tree/543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private static int calculateHeight(TreeNode root, int[] diameter) {
        if(root == null) return 0;
        
        int leftHeight = calculateHeight(root.left, diameter);
        int rightHeight = calculateHeight(root.right, diameter);
        
        diameter[0] = Math.max(diameter[0], leftHeight + rightHeight);
        return 1 + Math.max(leftHeight, rightHeight);
    }
    public int diameterOfBinaryTree(TreeNode root) {
        int diameter[] = new int[1];
        calculateHeight(root, diameter);
        return diameter[0];
    }
}